Sources/App/Core/SearchFilter/Filters/PlatformSearchFilter.swift

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020-2021 Dave Verwer, Sven A. Schmidt, and other contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import SQLKit
+
+/// Filters by ensuring the package is compatible with the platform provided.
+///
+/// Examples:
+/// ```
+/// platform:ios  - The package is compatible with iOS
+/// platform:!ios - The package is not compatible with iOS
+/// ```
+struct PlatformSearchFilter: SearchFilter {
+    static var key: String = "platform"
+    
+    var comparison: SearchFilterComparison
+    var value: Platform.Name
+    
+    init(value: String, comparison: SearchFilterComparison) throws {
+        guard [.match, .negativeMatch].contains(comparison) else {
+            throw SearchFilterError.unsupportedComparisonMethod
+        }
+        
+        guard let platform = Platform.Name(rawValue: value) else {
+            throw SearchFilterError.invalidValueType
+        }
+        
+        self.comparison = comparison
+        self.value = platform
+    }
+    
+    func `where`(_ builder: SQLPredicateGroupBuilder) -> SQLPredicateGroupBuilder {
+        let exp = SQLBinaryExpression(
+            left: SQLIdentifier("platforms").cast(to: "jsonb"),
+            op: SQLRaw("@>"),
+            right: SQLBind("[{\"name\":\"\(value.rawValue)\"}]").cast(to: "jsonb")
+        )
```

Comment:
Okay so the query I got working was this:

```
SELECT sp, package_name
FROM "versions", array_to_json(supported_platforms) AS sp 
WHERE sp::jsonb @> '[{"name":"ios"}]'::jsonb
LIMIT 100;
```

This shows all versions which supports iOS (or so my testing makes me believe).

Unfortunately SQL Kit does not make it easy to deal with:
1. JSON contains operators
2. JSONB casting (which is required to make this work)
3. JSON array syntax
4. Inverted (NOT) group expressions

I've tried to write my own helpers to make this a bit more readable but... 🤷

Reply:
Yes, fluent doesn't support JSON specific operators etc yet.

I haven't tried or looked this up but something like this might/should work

```swift
        let exp: SQLBinaryExpression = SQLRaw(
            #"platforms::JSONB @> '[{"name":\#(bind: value.rawValue)'::JSONB"#
        )
```

As the query itself, it might be easier to unpack the names from the supported platforms array so that essentially you're testing against a `text[]` of platform names. That should also open up using Fluent again.

Just a quick comment, I haven't looked into this very thoroughly.

Reply:
> it might be easier to unpack the names from the supported platforms array

I assume what you mean here is that `platforms` should be a text array which is `["ios","macos"]` instead of `[{"name": "tvos", "version": "10.0"},{"name": "ios", "version": "10.0"}]`? This was actually my first idea, I just couldn't figure out what the SQL query looked like to make that work - is there a specific function(s) I should be looking at?

Reply:
Just to add, I did have a play around with your example code above to tidy up the query.  

It looks like you have to use `SQLQueryString` to get the binding interpolation, but this doesn't actually work since the SQLBind feature wraps the binding in apostrophes (`'ios'`) which isn't valid JSON. It needs to be `"ios"` for the query to work 😞

Reply:
I just realised `supported_platforms` is the wrong field anyway:

```swift
    // TODO: rename to minimumPlatformVersions?
    @Field(key: "supported_platforms")
    var supportedPlatforms: [Platform]
```

which is filled from

```swift
    version.supportedPlatforms = manifest.platforms?.compactMap(Platform.init(from:)) ?? []
```

i.e. the manifests minimum required platform version field. Our version compatibility is from the build records (1-N on a version). That's not going to be searchable without some intermediate aggregation, either into a materialised view or into the version.

I don't think this search (not Swift version support) is doable without that (nontrivial) groundwork.

Reply:
That explains a lot about why so many packages didn't have data I suppose, since it's coming from a different source. Had my suspicion but didn't see the correlation with the Build model.

Will close this PR then, completely off the mark and need to come back at it differently.